[flexible-ipam] Update doc for Antrea 1.5

gran-vmv opened the pull request:
Signed-off-by: gran <gran@vmware.com>

jianjuns reviewed (commented):
We should move detailed description of a feature to a separate doc. feature-gates.md is only for brief description of the feature gate.

For IPAM, probably add the description to docs/antrea-ipam.md

gran-vmv commented:
> We should move detailed description of a feature to a separate doc. feature-gates.md is only for brief description of the feature gate.
> 
> For IPAM, probably add the description to docs/antrea-ipam.md

Moved detailed description to antrea-ipam.md and simplified feature-gates.md.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,131 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+## Antrea Flexible-IPAM
+
+`AntreaIPAM` is a feature that allows flexible control over Pod IP Addressing. This can
```
Probably we should say "Since version 1.4, Antrea starts to support flexible control over Pod IP addressing".

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
````diff
@@ -46,3 +46,131 @@ cluster. Valid range is 16 to 30. Default is 24.
+    ipam.antrea.io/ippools: 'pool1'
+```
+
+### Data path change for this feature
````
How about changing the title to "Data path behaviors"?

gran-vmv replied:
Sounds good. Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,131 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+### Requirements for this Feature
+
+In Antrea 1.4, this feature is supported on Linux Nodes, with IPv4, `system` OVS datapath
```
Remove "In Antrea 1.4". We can add the version info when a new version removes these limitations. Or you can say "As of now", "At this moment", etc.

gran-vmv replied:
Fixed. Thanks.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,131 @@ cluster. Valid range is 16 to 30. Default is 24.
+pool can be included in the Namespace annotation. In the future, annotation of up to two
+pools for IPv4 and IPv6 respectively will be supported.
+
+### Updates in Antrea 1.5
```
We should just talk about all features together, and indicate some features are available from which release; rather than creating a separate section for each release. We should make users easy to see all latest features.

gran-vmv replied:
Re-structured this part.

jianjuns replied:
Maybe we can add a section about "Usage", and put all IPPool, Namespace, Pod, Deployment, and StatefulSet examples there.

gran-vmv replied:
Fixed.

gran-vmv commented on docs/feature-gates.md:
```diff
@@ -207,7 +207,7 @@ This feature requires the Node IPAM Controller to be disabled in Kubernetes
+### Antrea Flexible-IPAM
 
 `AntreaIPAM` is a feature that allows flexible control over Pod IP Addressing. This can
 be achieved by configuring custom resource `IPPool` with desired set of IP ranges.
```
Fixed.

gran-vmv commented on docs/antrea-ipam.md:
````diff
@@ -46,3 +46,131 @@ cluster. Valid range is 16 to 30. Default is 24.
+      AntreaIPAM: true
+```
+
+Usage example:
````
Fixed.

gran-vmv commented on docs/feature-gates.md:
```diff
@@ -207,7 +207,7 @@ This feature requires the Node IPAM Controller to be disabled in Kubernetes
 there is a risk of conflicts in CIDR allocation between the two.
 
-### AntreaIPAM
+### Antrea Flexible-IPAM
```
We already have the usage section in `antrea-ipam-md`. Do we still need this in `feature-gates.md`?

jianjuns replied:
Sorry, I added the comment in wrong place. I actually meant antrea-ipam.md. I removed it here and moved it to antrea-ipam.md.

gran-vmv replied:
Got it.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+## Antrea Flexible-IPAM
+
+Since version 1.4, Antrea starts to support flexible control over Pod IP addressing. This
```
How about "Antrea supports flexible control over Pod IP addressing since version 1.4."

gran-vmv replied:
Sounds good. Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+## Antrea Flexible-IPAM
+
+Since version 1.4, Antrea starts to support flexible control over Pod IP addressing. This
+can be achieved by configuring custom resource `IPPool` with desired set of IP ranges.
```
Probably say: `IPPool` CRD

a desired set of

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+Since version 1.4, Antrea starts to support flexible control over Pod IP addressing. This
+can be achieved by configuring custom resource `IPPool` with desired set of IP ranges.
+The pool can be annotated to Namespace, Pod and PodTemplate of StatefulSet/Deployment.
```
pool -> `IPPool`

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+Since version 1.4, Antrea starts to support flexible control over Pod IP addressing. This
+can be achieved by configuring custom resource `IPPool` with desired set of IP ranges.
+The pool can be annotated to Namespace, Pod and PodTemplate of StatefulSet/Deployment.
+Antrea will manage IP address assignment for corresponding Pods according to pool spec.
```
the `IPPool` spec.

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+The pool can be annotated to Namespace, Pod and PodTemplate of StatefulSet/Deployment.
+Antrea will manage IP address assignment for corresponding Pods according to pool spec.
+
+Note that IP pool annotation cannot be updated, but rather re-created. IP Pool can be
```
IP pool annotation -> the IP pool annotation

What does "re-created" mean?

IP Pool -> "An IP pool" or "An IPPool"

gran-vmv replied:
Fixed.
```
Note that the IP pool annotation cannot be updated or deleted without recreating the
resource. An `IPPool` can be extended, ...
```

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+Antrea will manage IP address assignment for corresponding Pods according to pool spec.
+
+Note that IP pool annotation cannot be updated, but rather re-created. IP Pool can be
+extended, but cannot be shrunk if already assigned to a resource. The IP ranges of IP
```
"IP pools" or "IPPools"

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+extended, but cannot be shrunk if already assigned to a resource. The IP ranges of IP
+Pools must not overlap, otherwise it would lead to undefined behavior.
+
+Regular `Subnet per Node` IPAM will continue to be used for resources without IP pool
```
the IP pool annotation

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+#### IPPool Annotations on Pod (available since Antrea 1.5)
+
+In Antrea 1.5, Pod IPPool annotation is supported and have higher priority than Namespace
```
has a higher priority then the Namespace IPPool annotation

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+IPPool annotation. This annotation can be added to `PodTemplate` of a controller resource
+such as StatefulSet and Deployment.
+
+Pod IP annotation is supported for single Pod to specify a fixed IP for a Pod.
```
a single Pod

for the Pod

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+Pod IP annotation is supported for single Pod to specify a fixed IP for a Pod.
+
+Below 4 types are supported:
```
Probably say: "Examples of annotations on a Pod or PodTemplate“

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+#### Persistent IP for StatefulSet Pod (available since Antrea 1.5)
+
+In Antrea 1.5, StatefulSet Pod's IP will be kept when Pod restarts.
```
Remove "In Antrea 1.5"

A StatefulSet Pod's IP will be kept when Pod restarts, when the IP is allocated from the annotated IPPool.

gran-vmv replied:
Fixed. Thanks.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,145 @@ cluster. Valid range is 16 to 30. Default is 24.
+metadata:
+  annotations:
+    ipam.antrea.io/ippools: 'pod-ip-pool1'
+    ipam.antrea.io/pod-ips: '<ip-in-pod-ip-pool1>'
```
Should we use "ipam.antrea.io/podips"? What is our naming standard of annotations and labels? @tnqn

jianjuns replied:
Ok. Seems we do use "-" to connect words in annotations. Please ignore this comment.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,148 @@ cluster. Valid range is 16 to 30. Default is 24.
+## Antrea Flexible-IPAM
+
+Antrea supports flexible control over Pod IP addressing since version 1.4. This can be
+achieved by configuring custom resource `IPPool` CRD with a desired set of IP ranges.
```
Remove "custom resource".

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,148 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+#### Persistent IP for StatefulSet Pod (available since Antrea 1.5)
+
+A StatefulSet Pod's IP will be kept when Pod restarts, when the IP is allocated from the
```
Maybe change the first "when" to "after" to avoid two "when"s.

gran-vmv replied:
Fixed.

jianjuns commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,148 @@ cluster. Valid range is 16 to 30. Default is 24.
+
+#### IPPool Annotations on Pod (available since Antrea 1.5)
+
+In Antrea 1.5, Pod IPPool annotation is supported and have a higher priority than the
```
have -> has

jianjuns replied:
In -> Since

gran-vmv replied:
Fixed.

jianjuns reviewed (commented):
Several more nits.

jianjuns commented on docs/feature-gates.md:
```diff
@@ -209,59 +209,21 @@ there is a risk of conflicts in CIDR allocation between the two.
-corresponding Pods according to pool spec. In the future, support will be extended to
-Deployments and StatefulSets, as well as requesting specific IP via Pod annotation.
+`AntreaIPAM` feature allows flexible control over Pod IP addressing. This can be achieved
+by configuring custom resource `IPPool` with desired set of IP ranges. The pool can be
```
Could you also update this doc with the new revisions in antrea-ipam.md?

gran-vmv replied:
Sure. Synced the latest version to feature-gates.md

tnqn commented on docs/antrea-ipam.md:
```diff
@@ -46,3 +46,147 @@ cluster. Valid range is 16 to 30. Default is 24.
+    metadata:
+      annotations:
+        ipam.antrea.io/ippools: 'sts-ip-pool1'  # This annotation will be set automatically on all Pods managed by this resource
+        # Do not add pod-ips annotation to PodTemplate
```
Would the following be more accurate?
```suggestion
        # Do not add pod-ips annotation to PodTemplate if there is more than 1 replica
```

gran-vmv replied:
I think it is accurate, but I also feel not to confuse user to add `pod-ips` annotation into PodTemplate. If user increases replicas after StatefulSet with `pod-ips` PodTemplate annotation created, the result will be wrong.
Thus I prefer to notify user not to add `pod-ips` annotation to PodTemplate. For your use case, user should create a dedicated IPPool which contains >=1 free IP(s) for this StatefulSet.

tnqn replied:
But there is no such examples in this document yet. And if only 1 replica is desired for a StatefulSet or Deployment, it should be easier to just reuse the ippool to get a fixed IP?
Isn't the condition "if there is more than 1 replica" already clear that user shouldn't increase replica?

gran-vmv replied:
Got it. Added example for replicas=1 and replicas>1.

tnqn reviewed (commented):
LGTM, one minor comment

tnqn commented on docs/antrea-ipam.md:
````diff
@@ -46,3 +46,160 @@ cluster. Valid range is 16 to 30. Default is 24.
+```yaml
+kind: StatefulSet
+spec:
+  replicas: 1  # Do increase replicas if there is pod-ips annotation in PodTemplate
````
```suggestion
  replicas: 1  # Do not increase replicas if there is pod-ips annotation in PodTemplate
```

gran-vmv replied:
Fixed.

tnqn commented on docs/feature-gates.md:
```diff
@@ -209,59 +209,22 @@ there is a risk of conflicts in CIDR allocation between the two.
+behavior.
 
-Regular `Subnet per Node` IPAM will continue to be used for resources without IP pool
+Regular `Subnet per Node` IPAM will continue to be used for resources without the IP pool
```
As they are duplicate with the doc in antrea-ipam.md, can we remove the 2nd and 3rd paragraphs here and add a link in the end of the first paragraph like other features? "Refer to this [document](url) for more information."

gran-vmv replied:
Sure. Fixed.

tnqn commented:
/skip-all
